c10/util/C++17.h

```diff
@@ -123,10 +127,7 @@ constexpr auto apply_impl(F&& f, Tuple&& t, std::index_sequence<INDEX...>)
 }  // namespace detail
 
 template <class F, class Tuple>
-#if defined(_MSC_VER)
-C10_HOST_DEVICE // rocm doesn't like the C10_HOST_DEVICE
-#endif
-constexpr auto apply(F&& f, Tuple&& t) -> decltype(detail::apply_impl(
+CUDA_HOST_DEVICE constexpr auto apply(F&& f, Tuple&& t) -> decltype(detail::apply_impl(
```

Comment:
hm, it looks like before __host__ __device__ was here for windows only, and now it's for nvcc in general? I guess if it builds it's ok ;-)

Reply:
Yes, without this the compiler is compiling about calling a host function in device-host function is not allowed. I don't know the reason.

Reply:
It's probably better to change the `C10_HOST_DEVICE` macro to be disabled for rocm instead of introducing a new `CUDA_HOST_DEVICE`.

Reply:
@smessmer I know this is ugly, but

`C10_HOST_DEVICE` means to be 
if cpu --> ` ` (nothing),   if CUDA or ROCm --> `__device__ __host__`

and
`CUDA_HOST_DEVICE` means to be
if cpu or ROCm -->` ` (nothing),   if CUDA --> `__device__ __host__`

For most of the time in general, we should use `C10_HOST_DEVICE`, but the function in `C++17.h` belong to a case where CUDA requires the `__device__ __host__` qualifier while  ROCm reject it.

Reply:
ok then let's keep the current approach